src/EllCrv/NewtonLift.jl

```diff
@@ -0,0 +1,295 @@
+end
+
+function _compute_fn(point, ainvs)
+  P = parent(point[1])
+  @assert all(P===parent(a) for a in ainvs)
+  @assert all(P===parent(a) for a in point)
+  # x = xn//dn^2
+  # y = tn//dn^3
+  xn,yn,dn = point
+  a1,a2,a3,a4,a6 = ainvs
+  return yn^2 + a1*dn*xn*yn + a3*dn^3*yn - (xn^3 + a2*dn^2*xn^2 + a4*dn^4*xn + a6*dn^6)
+end
+
+function _newton_step(Rt, ainvsOKt, point, extra_points, prec)
+  @assert parent(point[1]) === parent(ainvsOKt[1])
+  # increase precision to the desired output precision
+  point = [setprecision!(i, 2*prec) for i in point]
+
+  fn = _compute_fn(point, ainvsOKt)
+
+  # check that the input has the desired precision
+  @assert minimum(_valuation(i) for i in coefficients(fn);init=inf)>=prec
+  @assert minimum(precision(i) for i in coefficients(fn); init=inf)>=prec
+  @assert all(minimum(precision(i) for i in coefficients(a); init=inf)>=2*prec for a in ainvsOKt)
+  @vprint :NewtonIteration 3 "input quality: $(prec)\n"
+  @assert all(minimum(precision(i) for i in coefficients(a); init=inf)>=2*prec for a in point)
+
+  R = base_ring(Rt)
+  OK = base_ring(R)
+  xn, yn, dn = point
+  OKt = parent(xn)
+  ainvsRt = [a(gen(Rt)) for a in ainvsOKt]
+  a1,a2,a3,a4,a6 = ainvsRt
+  t = gen(Rt)
+  xd = degree(xn)
+  yd = degree(yn)
+  dd = degree(dn)
+
+  xnt = xn(t)
+  ynt = yn(t)
+  dnt = dn(t)
+
+  xt = sum(gen(R,i+1)*t^i for i in 0:xd; init = zero(Rt))
+  yt = sum(gen(R, i+xd+2)*t^i for i in 0:yd; init = zero(Rt))
+  dt = sum(gen(R, i+xd+1+yd+2)*t^i for i in 0:dd-1; init = zero(Rt))#+t^dd
+
+  DF = (2*ynt + a1*dnt*xnt + a3*dnt^3)*yt +(a1*dnt*ynt- (3*xnt^2 + a2*dnt^2*2*xnt + a4*dnt^4))*xt + (a1*xnt*ynt + a3*3*dnt^2*ynt - (a2*2*dnt*xnt^2 + a4*4*dnt^3*xnt + a6*6*dnt^5))*dt
+  eqn = fn(t) + DF
+
+  equations = []
+  push!(equations, eqn)
+
+  EQ0 = elem_type(R)[]
+  EQc0 = elem_type(OK)[]
+  for (x0,y0,t0) in extra_points
+    eqn0 = xn(t0) + xt(t0) - (dn(t0)^2+2*dt*dn(t0))*x0
+    push!(equations, eqn0)
+    eqn0 = yn(t0) + yt(t0) -(dn(t0)^3+3*dt*dn(t0)^2)*y0
+    push!(equations, eqn0)
+  end
+
+  EQ = reduce(append!,[elem_type(R)[coeff(eqn, i) for i in 0:degree(eqn) if coeff(eqn, i)!=0] for eqn in equations])
+  EQc = reduce(append!, [elem_type(OK)[constant_coefficient(coeff(eqn, i)) for i in 0:degree(eqn) if coeff(eqn, i)!=0] for eqn in equations])
+  n = ngens(R) - 3
+
+  # each equation is a row
+  M = zero_matrix(OK, length(EQ), n)
+  for i in 1:length(EQ)
+    for j in 1:n
+      M[i,j] = coeff(EQ[i],gen(R,j))
+    end
+  end
+
+
+  EQc = -matrix(OK, length(EQc), 1, EQc)
+  fl, v = can_solve_with_solution_mod(M, EQc, 2*prec; side=:right)
+
+  fl || return false, point
+
+  #@show  minimum(_valuation.(M*v - EQc))
+  @assert minimum(_valuation.(M*v - EQc)) >= 2*prec
+  @assert minimum(_valuation.(v)) >= prec
+  t = gen(OKt)
+
+  xne = sum(v[i+1]*t^i for i in 0:xd; init=zero(OKt))
+  yne = sum(v[i+xd+2]*t^i for i in 0:yd; init=zero(OKt))
```

Comment:
Really? Computing powers of `t` this way? I agree that in this context this is not a runtime problem, but from a complexity point of view, you trade exactly `yd` multiplications with definitely more, possibly up to `yd^2/2`.
Assuming that `OKt` is a poly ring, a univariate one, then `OKt(v[xd+2:xd+yd])` or s.th. very similar, shuold do the job with 0 multiplications....

Reply:
Right, thanks for the reminder. Changed accordingly.